File: android/src/main/java/io/rong/flutter/imlib/MessageFactory.java
package io.rong.flutter.imlib;


import android.net.Uri;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.rong.imlib.model.ChatRoomInfo;
import io.rong.imlib.model.ChatRoomMemberInfo;
import io.rong.imlib.model.Conversation;
import io.rong.imlib.model.ConversationTagInfo;
import io.rong.imlib.model.MentionedInfo;
import io.rong.imlib.model.Message;
import io.rong.imlib.model.MessageConfig;
import io.rong.imlib.model.MessageContent;
import io.rong.imlib.model.MessagePushConfig;
import io.rong.imlib.model.ReadReceiptInfo;
import io.rong.imlib.model.SearchConversationResult;
import io.rong.imlib.model.TagInfo;
import io.rong.imlib.model.UserInfo;
import io.rong.imlib.typingmessage.TypingStatus;
import io.rong.message.GIFMessage;
import io.rong.message.ImageMessage;
import io.rong.message.ReferenceMessage;
import io.rong.message.SightMessage;
import io.rong.message.TextMessage;

public class MessageFactory {

    private static class SingleHolder {
        static MessageFactory instance = new MessageFactory();
    }

    public static MessageFactory getInstance() {
        return SingleHolder.instance;
    }

    public String message2String(Message message) {
        if (message == null) {
            return "";
        }

        Map map = messageToMap(message);

        JSONObject jObj = new JSONObject(map);

        String jStr = jObj.toString();

        return jStr;
    }

    public Map messageToMap(Message message) {
        Map map = new HashMap();
        map.put("conversationType", message.getConversationType().getValue());
        map.put("targetId", message.getTargetId());
        map.put("messageId", message.getMessageId());
        map.put("channelId", message.getChannelId());
        if (message.getMessageDirection() != null) {
            map.put("messageDirection", message.getMessageDirection().getValue());
        }
        map.put("senderUserId", message.getSenderUserId());
        if (message.getReceivedStatus() != null) {
            map.put("receivedStatus", message.getReceivedStatus().getFlag());
        }
        if (message.getSentStatus() != null) {
            map.put("sentStatus", message.getSentStatus().getValue());
        }
        ReadReceiptInfo readInfo = message.getReadReceiptInfo();
        if (readInfo != null) {
            HashMap readReceiptMap = new HashMap();
            readReceiptMap.put("isReceiptRequestMessage", readInfo.isReadReceiptMessage());
            readReceiptMap.put("hasRespond", readInfo.hasRespond());
            readReceiptMap.put("userIdList", readInfo.getRespondUserIdList());
            map.put("readReceiptInfo", readReceiptMap);
        }
        MessageConfig messageConfig = message.getMessageConfig();
        if (messageConfig != null) {
            HashMap messageConfigMap = new HashMap();
            messageConfigMap.put("disableNotification", messageConfig.isDisableNotification());
            map.put("messageConfig", messageConfigMap);
        }
        MessagePushConfig messagePushConfig = message.getMessagePushConfig();
        if (messagePushConfig != null) {
            HashMap messagePushConfigMap = new HashMap();
            messagePushConfigMap.put("pushTitle", messagePushConfig.getPushTitle());
            messagePushConfigMap.put("pushContent", messagePushConfig.getPushContent());
            messagePushConfigMap.put("pushData", messagePushConfig.getPushData());
            messagePushConfigMap.put("forceShowDetailContent", messagePushConfig.isForceShowDetailContent());
            messagePushConfigMap.put("disablePushTitle", messagePushConfig.isDisablePushTitle());
            messagePushConfigMap.put("templateId", messagePushConfig.getTemplateId());
            if (messagePushConfig.getAndroidConfig() != null) {
                HashMap androidConfigMap = new HashMap();
                androidConfigMap.put("notificationId", messagePushConfig.getAndroidConfig().getNotificationId());
                androidConfigMap.put("channelIdMi", messagePushConfig.getAndroidConfig().getChannelIdMi());
                androidConfigMap.put("channelIdHW", messagePushConfig.getAndroidConfig().getChannelIdHW());
                androidConfigMap.put("channelIdOPPO", messagePushConfig.getAndroidConfig().getChannelIdOPPO());
                androidConfigMap.put("typeVivo", messagePushConfig.getAndroidConfig().getTypeVivo());
                messagePushConfigMap.put("androidConfig", androidConfigMap);
            }
            if (messagePushConfig.getIOSConfig() != null) {
                HashMap iosConfigMap = new HashMap();
                iosConfigMap.put("thread_id", messagePushConfig.getIOSConfig().getThread_id());
                iosConfigMap.put("apns_collapse_id", messagePushConfig.getIOSConfig().getApns_collapse_id());
                messagePushConfigMap.put("iOSConfig", iosConfigMap);
            }
            map.put("messagePushConfig", messagePushConfigMap);
        }
        map.put("sentTime", message.getSentTime());
        map.put("objectName", message.getObjectName());
        map.put("extra", message.getExtra());
        map.put("canIncludeExpansion", message.isCanIncludeExpansion());
        map.put("expansionDic", message.getExpansion());
        String uid = message.getUId();
        if (uid == null || uid.length() <= 0) {
            uid = "";
        }
        map.put("messageUId", uid);

        MessageContent content = message.getContent();
        if (message.getContent() instanceof ImageMessage) {
            RCMessageHandler.encodeImageMessage(message);
        } else if (message.getContent() instanceof SightMessage) {
            RCMessageHandler.encodeSightMessage(message);
        } else if (message.getContent() instanceof GIFMessage) {
            RCMessageHandler.encodeGifMessage(message);
        } else if (message.getContent() instanceof ReferenceMessage) {
            // 引用消息的引用内容的类型需要判断
            RCMessageHandler.encodeReferenceMessage(message);
        }
        // 判断 TextMessage 内容不能为 null
        if (content instanceof TextMessage) {
            if (((TextMessage) content).getContent() == null) {
                ((TextMessage) content).setContent("");
            }
        }

        byte[] data = null;
        if (content instanceof ImageMessage) {
            // 处理 thumbUri 丢失的问题
            data = RCMessageHandler.encodeImageContent((ImageMessage) content);
        } else if (content instanceof SightMessage) {
            data = RCMessageHandler.encodeSightContent((SightMessage) content);
        } else if (content != null) {
            data = content.encode();
        }

        if (data != null && data.length > 0) {
            String jsonS = new String(data);
            map.put("content", jsonS);
        }
        return map;
    }

    public String messageContent2String(MessageContent content) {
        if (content == null) {
            return null;
        }
        byte[] data = content.encode();
        String jsonS = new String(data);
        return jsonS;
    }

    public String conversation2String(Conversation conversation) {
        Map map = new HashMap();
        map.put("conversationType", conversation.getConversationType().getValue());
        map.put("targetId", conversation.getTargetId());
        map.put("channelId", conversation.getChannelId());
        map.put("unreadMessageCount", conversation.getUnreadMessageCount());
        map.put("receivedStatus", conversation.getReceivedStatus().getFlag());
        map.put("sentStatus", conversation.getSentStatus().getValue());
        map.put("sentTime", conversation.getSentTime());
        map.put("isTop", conversation.isTop());
        map.put("objectName", conversation.getObjectName());
        map.put("senderUserId", conversation.getSenderUserId());
        map.put("latestMessageId", conversation.getLatestMessageId());
        map.put("mentionedCount", conversation.getMentionedCount());
        map.put("draft", conversation.getDraft());
        if (conversation.getNotificationStatus() != null) {
            map.put("blockStatus", conversation.getNotificationStatus().getValue());
        }
        map.put("receivedTime", conversation.getReceivedTime());

        MessageContent content = conversation.getLatestMessage();
        if (content != null) {
            byte[] data = content.encode();
            if (data != null && data.length > 0) {
                String jsonS = new String(data);
                map.put("content", jsonS);
            }
        } else {
//            map.put("content","");
        }


        JSONObject jObj = new JSONObject(map);

        String jStr = jObj.toString();
        return jStr;
    }

    public String conversationTagInfo2String(ConversationTagInfo conversationTagInfo) {
        if (conversationTagInfo == null) {
            return "";
        }
        Map map = new HashMap();
        map.put("tagInfo", tagInfo2String(conversationTagInfo.getTagInfo()));
        map.put("isTop", conversationTagInfo.isTop());
        JSONObject jObj = new JSONObject(map);

        String jStr = jObj.toString();
        return jStr;
    }

    public String tagInfo2String(TagInfo tagInfo) {
        if (tagInfo == null) {
            return "";
        }
        Map map = new HashMap();
        map.put("tagId", tagInfo.getTagId());
        map.put("tagName", tagInfo.getTagName());
        map.put("count", tagInfo.getCount());
        map.put("timestamp", tagInfo.getTimestamp());

        JSONObject jObj = new JSONObject(map);

        String jStr = jObj.toString();
        return jStr;
    }

    public Map chatRoom2Map(ChatRoomInfo chatRoomInfo) {
        Map map = new HashMap();
        map.put("targetId", chatRoomInfo.getChatRoomId());
        map.put("memberOrder", chatRoomInfo.getMemberOrder().getValue());
        map.put("totalMemeberCount", chatRoomInfo.getTotalMemberCount());

        List memList = new ArrayList();
        for (ChatRoomMemberInfo memInfo : chatRoomInfo.getMemberInfo()) {
            Map memMap = new HashMap();
            memMap.put("userId", memInfo.getUserId());
            memMap.put("joinTime", memInfo.getJoinTime());
            memList.add(memMap);
        }
        map.put("memberInfoList", memList);

        return map;
    }

    public String typingStatus2String(TypingStatus status) {
        Map map = new HashMap();
        map.put("userId", status.getUserId());
        map.put("typingContentType", status.getTypingContentType());
        map.put("sentTime", status.getSentTime());
        JSONObject jObj = new JSONObject(map);
        String jStr = jObj.toString();
        return jStr;
    }

    public String SearchConversationResult2String(SearchConversationResult result) {
        Map map = new HashMap();
        map.put("mConversation", conversation2String(result.getConversation()));
        map.put("mMatchCount", result.getMatchCount());
        JSONObject jObj = new JSONObject(map);
        String jStr = jObj.toString();
        return jStr;
    }

    public void setCommonInfo(String contentStr, MessageContent content) {
        try {
            JSONObject contentObject = new JSONObject(contentStr);
            if (contentObject.has("user")) {
                Object userObject = contentObject.get("user");
                if (userObject instanceof JSONObject) {
                    JSONObject userJObject = (JSONObject) userObject;
                    String id = "";
                    String name = "";
                    String portrait = "";
                    if (userJObject.has("id")) {
                        id = (String) userJObject.get("id");
                    }
                    if (userJObject.has("name")) {
                        name = (String) userJObject.get("name");
                    }
                    if (userJObject.has("portrait")) {
                        portrait = (String) userJObject.get("portrait");
                    }
                    UserInfo info = new UserInfo(id, name, Uri.parse(portrait));
                    if (userJObject.has("extra")) {
                        info.setExtra((String) userJObject.get("extra"));
                    }
                    content.setUserInfo(info);
                }
            }
            if (contentObject.has("mentionedInfo")) {
                Object mentionedObject = contentObject.get("mentionedInfo");
                if (mentionedObject instanceof JSONObject) {
                    JSONObject mentionedJObject = (JSONObject) mentionedObject;
                    MentionedInfo info = new MentionedInfo();
                    if (mentionedJObject.has("type")) {
                        info.setType(MentionedInfo.MentionedType.valueOf((int) mentionedJObject.get("type")));
                    }
                    if (mentionedJObject.has("userIdList")) {
                        JSONArray userIdArray = (JSONArray) mentionedJObject.get("userIdList");
                        List<String> userIdList = new ArrayList<>();
                        for (int i = 0; i < userIdArray.length(); i++) {
                            String idStr = (String) userIdArray.get(i);
                            userIdList.add(idStr);
                        }
                        info.setMentionedUserIdList(userIdList);
                    }
                    if (mentionedJObject.has("mentionedContent")) {
                        info.setMentionedContent((String) mentionedJObject.get("mentionedContent"));
                    }
                    content.setMentionedInfo(info);
                }
            }
            if (contentObject.has("burnDuration")) {
                long burnDuration = Long.valueOf(contentObject.get("burnDuration").toString());
                content.setDestructTime(burnDuration);
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
    }
}

File: lib/src/util/message_factory.dart
import 'dart:convert' show json;
import 'dart:core';
import 'dart:developer' as developer;

import 'package:rongcloud_im_plugin/src/info/tag_info.dart';

import '../../rongcloud_im_plugin.dart';
import '../util/type_util.dart';

class MessageFactory extends Object {
  factory MessageFactory() => _getInstance()!;

  static MessageFactory? get instance => _getInstance();
  static MessageFactory? _instance;

  MessageFactory._internal() {
    // 初始化
  }

  static MessageFactory? _getInstance() {
    if (_instance == null) {
      _instance = new MessageFactory._internal();
    }
    return _instance;
  }

  ConversationTagInfo? string2ConversationTagInfo(String tagInfoJsonStr) {
    if (TypeUtil.isEmptyString(tagInfoJsonStr)) {
      return null;
    }
    Map map = json.decode(tagInfoJsonStr);
    return map2ConversationTagInfo(map);
  }

  ConversationTagInfo map2ConversationTagInfo(Map map) {
    ConversationTagInfo conversationTagInfo = new ConversationTagInfo();
    conversationTagInfo.isTop = map["isTop"];
    String? jsonTagStr = map["tagInfo"];
    conversationTagInfo.tagInfo = string2TagInfo(jsonTagStr);
    return conversationTagInfo;
  }

  TagInfo? string2TagInfo(String? tagJsonStr) {
    if (TypeUtil.isEmptyString(tagJsonStr)) {
      return null;
    }
    Map map = json.decode(tagJsonStr!);
    return map2TagInfo(map);
  }

  TagInfo map2TagInfo(Map map) {
    TagInfo tagInfo = new TagInfo();
    tagInfo.tagId = map["tagId"];
    tagInfo.tagName = map["tagName"];
    tagInfo.count = map["count"];
    tagInfo.timestamp = map["timestamp"];
    return tagInfo;
  }

  Message? string2Message(String? msgJsonStr) {
    if (TypeUtil.isEmptyString(msgJsonStr)) {
      return null;
    }
    Map map = json.decode(msgJsonStr!);
    return map2Message(map);
  }

  Conversation? string2Conversation(String? conJsonStr) {
    if (TypeUtil.isEmptyString(conJsonStr)) {
      return null;
    }
    Map map = json.decode(conJsonStr!);
    return map2Conversation(map);
  }

  ChatRoomInfo map2ChatRoomInfo(Map map) {
    ChatRoomInfo chatRoomInfo = new ChatRoomInfo();
    chatRoomInfo.targetId = map["targetId"];
    chatRoomInfo.memberOrder = map["memberOrder"];
    chatRoomInfo.totalMemeberCount = map["totalMemeberCount"];
    List memList = [];
    for (Map memMap in map["memberInfoList"]) {
      memList.add(map2ChatRoomMemberInfo(memMap));
    }
    chatRoomInfo.memberInfoList = memList;
    return chatRoomInfo;
  }

  ChatRoomMemberInfo map2ChatRoomMemberInfo(Map map) {
    ChatRoomMemberInfo chatRoomMemberInfo = new ChatRoomMemberInfo();
    chatRoomMemberInfo.userId = map["userId"];
    chatRoomMemberInfo.joinTime = map["joinTime"];
    return chatRoomMemberInfo;
  }

  Message map2Message(Map map) {
    Message message = new Message();
    message.conversationType = map["conversationType"];
    message.targetId = map["targetId"];
    message.messageId = map["messageId"];
    message.messageDirection = map["messageDirection"];
    message.senderUserId = map["senderUserId"];
    message.receivedStatus = map["receivedStatus"];
    message.sentStatus = map["sentStatus"];
    message.sentTime = map["sentTime"];
    message.objectName = map["objectName"];
    message.messageUId = map["messageUId"];
    message.extra = map["extra"];
    message.canIncludeExpansion = map["canIncludeExpansion"];
    message.expansionDic = map["expansionDic"];
    Map? messageConfigMap = map["messageConfig"];
    if (messageConfigMap != null) {
      MessageConfig messageConfig = MessageConfig();
      messageConfig.disableNotification = messageConfigMap["disableNotification"];
      message.messageConfig = messageConfig;
    }
    Map? messagePushConfigMap = map["messagePushConfig"];
    if (messagePushConfigMap != null) {
      message.messagePushConfig = MessagePushConfig();
      message.messagePushConfig!.pushTitle = messagePushConfigMap["pushTitle"];
      message.messagePushConfig!.pushContent = messagePushConfigMap["pushContent"];
      message.messagePushConfig!.pushData = messagePushConfigMap["pushData"];
      message.messagePushConfig!.forceShowDetailContent = messagePushConfigMap["forceShowDetailContent"];
      Map? androidConfigMap = messagePushConfigMap["androidConfig"];
      if (androidConfigMap != null) {
        AndroidConfig androidConfig = AndroidConfig();
        androidConfig.notificationId = androidConfigMap["notificationId"];
        androidConfig.channelIdMi = androidConfigMap["channelIdMi"];
        androidConfig.channelIdHW = androidConfigMap["channelIdHW"];
        androidConfig.channelIdOPPO = androidConfigMap["channelIdOPPO"];
        androidConfig.typeVivo = androidConfigMap["typeVivo"];
        message.messagePushConfig!.androidConfig = androidConfig;
      }
      Map? iOSConfigMap = messagePushConfigMap["iOSConfig"];
      if (iOSConfigMap != null) {
        IOSConfig iosConfig = IOSConfig();
        iosConfig.thread_id = iOSConfigMap["thread_id"];
        iosConfig.apns_collapse_id = iOSConfigMap["apns_collapse_id"];
        message.messagePushConfig!.iOSConfig = iosConfig;
      }
      message.messagePushConfig!.disablePushTitle = messagePushConfigMap["disablePushTitle"];
      message.messagePushConfig!.templateId = messagePushConfigMap["templateId"];
    }
    Map? readReceiptMap = map["readReceiptInfo"];
    if (readReceiptMap != null) {
      ReadReceiptInfo readReceiptInfo = ReadReceiptInfo();
      readReceiptInfo.isReceiptRequestMessage = readReceiptMap["isReceiptRequestMessage"];
      readReceiptInfo.hasRespond = readReceiptMap["hasRespond"];
      readReceiptInfo.userIdList = readReceiptMap["userIdList"];
      message.readReceiptInfo = readReceiptInfo;
    }
    String? contenStr = map["content"];
    MessageContent? content = string2MessageContent(contenStr, message.objectName);
    if (contenStr == null || contenStr == "") {
      developer.log(message.objectName! + ":该消息内容为空，可能该消息没有在原生 SDK 中注册", name: "RongIMClient.MessageFactory");
      return message;
    }
    if (content != null) {
      message.content = content;
    } else {
      developer.log("${message.objectName}:该消息不能被解析!消息内容被保存在 Message.originContentMap 中",
          name: "RongIMClient.MessageFactory");
      Map? map = json.decode(contenStr.toString());
      message.originContentMap = map;
    }
    return message;
  }

  Conversation map2Conversation(Map map) {
    Conversation con = new Conversation();
    con.conversationType = map["conversationType"];
    con.targetId = map["targetId"];
    con.channelId = map["channelId"];
    con.unreadMessageCount = map["unreadMessageCount"];
    con.receivedStatus = map["receivedStatus"];
    con.sentStatus = map["sentStatus"];
    con.sentTime = map["sentTime"];
    con.isTop = map["isTop"];
    con.objectName = map["objectName"];
    con.senderUserId = map["senderUserId"];
    con.latestMessageId = map["latestMessageId"];
    con.mentionedCount = map["mentionedCount"];
    con.draft = map["draft"];
    con.blockStatus = map["blockStatus"];
    con.receivedTime = map["receivedTime"];

    String? contenStr = map["content"];
    MessageContent? content = string2MessageContent(contenStr, con.objectName);
    if (content != null) {
      con.latestMessageContent = content;
    } else {
      if (contenStr == null || contenStr.length <= 0) {
        developer.log("该会话没有消息 type:" + con.conversationType.toString() + " targetId:" + con.targetId!,
            name: "RongIMClient.MessageFactory");
      } else {
        developer.log(con.objectName! + ":该消息不能被解析!消息内容被保存在 Conversation.originContentMap 中",
            name: "RongIMClient.MessageFactory");
        Map? map = json.decode(contenStr.toString());
        con.originContentMap = map;
      }
    }
    return con;
  }

  MessageContent? string2MessageContent(String? contentS, String? objectName) {
    MessageContent? content;
    if (RongIMClient.messageDecoders.isNotEmpty && RongIMClient.messageDecoders.containsKey(objectName)) {
      return RongIMClient.messageDecoders[objectName!]!(contentS);
    }
    return content;
  }

  Map conversationIdentifier2Map(ConversationIdentifier? identifier) {
    Map map = new Map();
    if (identifier == null) {
      return map;
    }
    map["conversationType"] = identifier.conversationType;
    map["targetId"] = identifier.targetId;
    return map;
  }

  Map messageContent2Map(MessageContent content) {
    Map map = new Map();
    return map;
  }

  Map message2Map(Message message) {
    Map map = new Map();
    map["conversationType"] = message.conversationType;
    map["targetId"] = message.targetId ?? "";
    map["messageId"] = message.messageId;
    map["messageDirection"] = message.messageDirection;
    map["senderUserId"] = message.senderUserId ?? "";
    map["receivedStatus"] = message.receivedStatus;
    map["sentStatus"] = message.sentStatus;
    map["sentTime"] = message.sentTime ?? 0;
    map["objectName"] = message.objectName ?? "";
    map["messageUId"] = message.messageUId ?? "";
    if (message.content != null) {
      map["content"] = message.content!.encode();
    }
    map["extra"] = message.extra ?? "";
    map["canIncludeExpansion"] = message.canIncludeExpansion ?? false;
    map["expansionDic"] = message.expansionDic;
    if (message.messageConfig != null) {
      Map messageConfig = Map();
      messageConfig["disableNotification"] = message.messageConfig!.disableNotification ?? false;
      map["messageConfig"] = messageConfig;
    }
    if (message.messagePushConfig != null) {
      Map messagePushConfig = Map();
      messagePushConfig["pushTitle"] = message.messagePushConfig!.pushTitle ?? "";
      messagePushConfig["pushContent"] = message.messagePushConfig!.pushContent ?? "";
      messagePushConfig["pushData"] = message.messagePushConfig!.pushData ?? "";
      messagePushConfig["forceShowDetailContent"] = message.messagePushConfig!.forceShowDetailContent ?? false;
      messagePushConfig["disablePushTitle"] = message.messagePushConfig!.disablePushTitle ?? false;
      messagePushConfig["templateId"] = message.messagePushConfig!.templateId ?? "";
      if (message.messagePushConfig!.androidConfig != null) {
        Map androidConfig = Map();
        androidConfig["notificationId"] = message.messagePushConfig!.androidConfig!.notificationId ?? "";
        androidConfig["channelIdMi"] = message.messagePushConfig!.androidConfig!.channelIdMi ?? "";
        androidConfig["channelIdHW"] = message.messagePushConfig!.androidConfig!.channelIdHW ?? "";
        androidConfig["channelIdOPPO"] = message.messagePushConfig!.androidConfig!.channelIdOPPO ?? "";
        androidConfig["typeVivo"] = message.messagePushConfig!.androidConfig!.typeVivo ?? "";
        messagePushConfig["androidConfig"] = androidConfig;
      }
      if (message.messagePushConfig!.iOSConfig != null) {
        Map iOSConfig = Map();
        iOSConfig["thread_id"] = message.messagePushConfig!.iOSConfig!.thread_id ?? "";
        iOSConfig["apns_collapse_id"] = message.messagePushConfig!.iOSConfig!.apns_collapse_id ?? "";
        messagePushConfig["iOSConfig"] = iOSConfig;
      }
      map["messagePushConfig"] = messagePushConfig;
    }
    if (message.readReceiptInfo != null) {
      Map readReceiptMap = Map();
      readReceiptMap["isReceiptRequestMessage"] = message.readReceiptInfo!.isReceiptRequestMessage;
      readReceiptMap["hasRespond"] = message.readReceiptInfo!.hasRespond;
      readReceiptMap["userIdList"] = message.readReceiptInfo!.userIdList;
      map["readReceiptInfo"] = readReceiptMap;
    }

    return map;
  }

  TypingStatus? string2TypingStatus(String statusJsonStr) {
    if (TypeUtil.isEmptyString(statusJsonStr)) {
      return null;
    }
    Map map = json.decode(statusJsonStr);
    return map2TypingStatus(map);
  }

  TypingStatus map2TypingStatus(Map map) {
    TypingStatus status = new TypingStatus();
    status.userId = map["userId"];
    status.typingContentType = map["typingContentType"];
    status.sentTime = map["sentTime"];
    return status;
  }

  SearchConversationResult? string2SearchConversationResult(String resultStr) {
    if (TypeUtil.isEmptyString(resultStr)) {
      return null;
    }
    Map map = json.decode(resultStr);
    return map2SearchConversationResult(map);
  }

  SearchConversationResult map2SearchConversationResult(Map map) {
    SearchConversationResult result = new SearchConversationResult();
    result.mConversation = string2Conversation(map["mConversation"]);
    result.mMatchCount = map["mMatchCount"];
    return result;
  }

  Map pushConfig2Map(PushConfig pushConfig) {
    Map map = new Map();
    map["enableHWPush"] = pushConfig.enableHWPush;
    map["enableFCM"] = pushConfig.enableFCM;
    map["enableVivoPush"] = pushConfig.enableVivoPush;
    map["miAppId"] = pushConfig.miAppId;
    map["miAppKey"] = pushConfig.miAppKey;
    map["mzAppId"] = pushConfig.mzAppId;
    map["mzAppKey"] = pushConfig.mzAppKey;
    map["oppoAppKey"] = pushConfig.oppoAppKey;
    map["oppoAppSecret"] = pushConfig.oppoAppSecret;
    return map;
  }
}
